### PR TITLE
8280481: Duplicated static stubs in NMethod Stub Code section

### PR DESCRIPTION
**Pull request by eastig**
Calls of Java methods have stubs to the interpreter for the cases when an invoked Java method is not compiled. Calls of static Java methods and final Java methods have statically bound information about a callee during compilation. Such calls can share stubs to the interpreter.

Each stub to the interpreter has a relocation record (accessed via `relocInfo`) which provides an address of the stub and an address of its owner. `relocInfo` has an offset which is an offset from the previously know relocatable address. The address of a stub is calculated as the address provided by the previous `relocInfo` plus the offset.

Each Java call has:
- A relocation for a call site.
- A relocation for a stub to the interpreter.
- A stub to the interpreter.
- If far jumps are used (arm64 case):
  - A trampoline relocation.
  - A trampoline.

We cannot avoid creating relocations. They are needed to support patching call sites and stubs.

One approach to create shared stubs to keep track of created stubs. If the needed stub exist we use its address and create only needed relocation information. The `relocInfo` for a created stub will have a positive offset. As relocations for different stubs can be created after that, a relocation for a shared stub will have a negative offset relative to the address provided by the previous relocation:
```
reloc1  ---> 0x0: stub1
reloc2  ---> 0x4: stub2 (reloc2.addr = reloc1.addr + reloc2.offset = 0x0 + 4)
reloc3  ---> 0x0: stub1 (reloc3.addr = reloc2.addr + reloc3.offset = 0x4 - 4)
```
According to [relocInfo.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/relocInfo.hpp#L237):
```
// [About Offsets] Relative offsets are supplied to this module as
// positive byte offsets, but they may be internally stored scaled
// and/or negated, depending on what is most compact for the target
// system.  Since the object pointed to by the offset typically
// precedes the relocation address, it is profitable to store
// these negative offsets as positive numbers, but this decision
// is internal to the relocation information abstractions.
```
However, `CodeSection` does not support negative offsets. It [assumes](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/asm/codeBuffer.hpp#L195) addresses relocations pointing at grow upward:
```
class CodeSection {
...
 private:
...
    address     _locs_point;      // last relocated position (grows upward)
...
  void    set_locs_point(address pc) {
    assert(pc >= locs_point(), "relocation addr may not decrease");
    assert(allocates2(pc),     "relocation addr must be in this section");
    _locs_point = pc;
  }
```
Negative offsets reduce the offset range by half. This can cause the increase of filler records, the empty `relocInfo` records to reduce offset values. Also negative offsets are only needed for `static_stub_type`, but other 13 types don’t need them.

This PR implements another approach: postponed creation of stubs. First we collect requests for creating shared stubs. Then we have the finalisation phase, where shared stubs are created in `CodeBuffer`. This approach does not need negative offsets. Supported platforms are x86, x86_64 and aarch64.

There is a new diagnostic option: `UseSharedStubs`. Its default value for x86, x86_64 and aarch64 is set true.

**Results from [Renaissance 0.14.0](https://github.com/renaissance-benchmarks/renaissance/releases/tag/v0.14.0)**
Note: 'Nmethods with shared stubs' is the total number of nmethods counted during benchmark's run. 'Final # of nmethods' is a number of nmethods in CodeCache when JVM exited.
- AArch64
```
+------------------+-------------+----------------------------+---------------------+
|    Benchmark     | Saved bytes | Nmethods with shared stubs | Final # of nmethods |
+------------------+-------------+----------------------------+---------------------+
| dotty            |      820544 |                       4592 |               18872 |
| dec-tree         |      405280 |                       2580 |               22335 |
| naive-bayes      |      392384 |                       2586 |               21184 |
| log-regression   |      362208 |                       2450 |               20325 |
| als              |      306048 |                       2226 |               18161 |
| finagle-chirper  |      262304 |                       2087 |               12675 |
| movie-lens       |      250112 |                       1937 |               13617 |
| gauss-mix        |      173792 |                       1262 |               10304 |
| finagle-http     |      164320 |                       1392 |               11269 |
| page-rank        |      155424 |                       1175 |               10330 |
| chi-square       |      140384 |                       1028 |                9480 |
| akka-uct         |      115136 |                        541 |                3941 |
| reactors         |       43264 |                        335 |                2503 |
| scala-stm-bench7 |       42656 |                        326 |                3310 |
| philosophers     |       36576 |                        256 |                2902 |
| scala-doku       |       35008 |                        231 |                2695 |
| rx-scrabble      |       32416 |                        273 |                2789 |
| future-genetic   |       29408 |                        260 |                2339 |
| scrabble         |       27968 |                        225 |                2477 |
| par-mnemonics    |       19584 |                        168 |                1689 |
| fj-kmeans        |       19296 |                        156 |                1647 |
| scala-kmeans     |       18080 |                        140 |                1629 |
| mnemonics        |       17408 |                        143 |                1512 |
+------------------+-------------+----------------------------+---------------------+
```
- X86_64
```
+------------------+-------------+----------------------------+---------------------+
|    Benchmark     | Saved bytes | Nmethods with shared stubs | Final # of nmethods |
+------------------+-------------+----------------------------+---------------------+
| dotty            |      337065 |                       4403 |               19135 |
| dec-tree         |      183045 |                       2559 |               22071 |
| naive-bayes      |      176460 |                       2450 |               19782 |
| log-regression   |      162555 |                       2410 |               20648 |
| als              |      121275 |                       1980 |               17179 |
| movie-lens       |      111915 |                       1842 |               13020 |
| finagle-chirper  |      106350 |                       1947 |               12726 |
| gauss-mix        |       81975 |                       1251 |               10474 |
| finagle-http     |       80895 |                       1523 |               12294 |
| page-rank        |       68940 |                       1146 |               10124 |
| chi-square       |       62130 |                        974 |                9315 |
| akka-uct         |       50220 |                        555 |                4263 |
| reactors         |       23385 |                        371 |                2544 |
| philosophers     |       17625 |                        259 |                2865 |
| scala-stm-bench7 |       17235 |                        295 |                3230 |
| scala-doku       |       15600 |                        214 |                2698 |
| rx-scrabble      |       14190 |                        262 |                2770 |
| future-genetic   |       13155 |                        253 |                2318 |
| scrabble         |       12300 |                        217 |                2352 |
| fj-kmeans        |        8985 |                        157 |                1616 |
| par-mnemonics    |        8535 |                        155 |                1684 |
| scala-kmeans     |        8250 |                        138 |                1624 |
| mnemonics        |        7485 |                        134 |                1522 |
+------------------+-------------+----------------------------+---------------------+
```

**Testing: fastdebug and release builds for x86, x86_64 and aarch64**
- `tier1`...`tier4`: Passed
- `hotspot/jtreg/compiler/sharedstubs`: Passed

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8280481](https://bugs.openjdk.java.net/browse/JDK-8280481): Duplicated static stubs in NMethod Stub Code section

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8024/head:pull/8024` \
`$ git checkout pull/8024`

Update a local copy of the PR: \
`$ git checkout pull/8024` \
`$ git pull https://git.openjdk.java.net/jdk pull/8024/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8024`

View PR using the GUI difftool: \
`$ git pr show -t 8024`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8024.diff">https://git.openjdk.java.net/jdk/pull/8024.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/asm/codeBuffer.cpp [hunk 1]
```diff
@@ -979,6 +980,22 @@ void CodeBuffer::log_section_sizes(const char* name) {
...
+void CodeBuffer::finalize_stubs() {
+  if (!pd_finalize_stubs()) {
+    return;
+  }
+  _finalize_stubs = false;
+}
+
+void CodeBuffer::shared_stub_to_interp_for(Method* method, address caller_pc) {
+  if (_shared_stub_to_interp_requests == NULL) {
+    _shared_stub_to_interp_requests = new SharedStubToInterpRequests();
```
[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/share/asm/codeBuffer.cpp:992]: For the implementation of `SharedStubToInterpRequests` based on `GrowableArray`, there is no need to do this explicitly. By default `GrowableArray` is allocated in thread's resource area which should be automatically cleaned up with a proper set resource mark. A resource mark is set in [CompileBroker::invoke_compiler_on_method](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/compiler/compileBroker.cpp#L2166).

File: src/hotspot/share/runtime/globals.hpp [hunk 2]
```diff
@@ -2026,8 +2026,11 @@ const intx ObjectAlignmentInBytes = 8;
              "Mark all threads after a safepoint, and clear on a modify "   \
              "fence. Add cleanliness checks.")                              \
                                                                             \
-  develop(bool, TraceOptimizedUpcallStubs, false,                              \
-                "Trace optimized upcall stub generation")                      \
+  develop(bool, TraceOptimizedUpcallStubs, false,                           \
+                "Trace optimized upcall stub generation")                   \
+  product(bool, UseSharedStubs, false, DIAGNOSTIC,                          \
```
Comment (bulasevich) [inline src/hotspot/share/runtime/globals.hpp:2031]: Please don't add a new option. We already have a plenty of them.

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+ * @run driver compiler.sharedstubs.SharedStubToInterpTest c1 FinalClassTest
+ * @run driver compiler.sharedstubs.SharedStubToInterpTest c2 FinalMethodTest
+ * @run driver compiler.sharedstubs.SharedStubToInterpTest c1 FinalMethodTest
+ */
+
+package compiler.sharedstubs;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.ListIterator;
```
Comment (bulasevich) [inline test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java:45]: excessive import

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+/**
+ * @test SharedStubToInterpTest
+ * @summary Checks that stubs to the interpreter can be shared for static or final method.
+ * @bug 8280481
+ * @library /test/lib
+ *
+ * @requires os.arch=="amd64" | os.arch=="x86_64" | os.arch=="i386" | os.arch=="x86" | os.arch=="aarch64"
+ *
+ * @run driver compiler.sharedstubs.SharedStubToInterpTest c2 StaticMethodTest
```
Comment (bulasevich) [inline test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java:33]: Is not it better to run all the VM instances from a single test?
```java
List<String> options = java.util.Arrays.asList("-XX:-TieredCompilation", "-XX:TieredStopAtLevel=1");
List<String> tests = java.util.Arrays.asList("StaticMethodTest", "FinalClassTest", "FinalMethodTest");
for (String option: options) {
    for (String test: tests) {
        runVM(option, test);
    }
} 
```

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        command.add("-Xbatch");
+        command.add("-XX:CompileCommand=compileonly," + testClassName + "::" + "test");
+        command.add("-XX:CompileCommand=dontinline," + testClassName + "::" + "test");
+        command.add("-XX:CompileCommand=print," + testClassName + "::" + "test");
+        command.add("-XX:CompileCommand=exclude," + testClassName + "::" + "log01");
+        command.add("-XX:CompileCommand=dontinline," + testClassName + "::" + "log01");
+        command.add("-XX:CompileCommand=exclude," + testClassName + "::" + "log02");
+        command.add("-XX:CompileCommand=dontinline," + testClassName + "::" + "log02");
+        command.add(testClassName);
+        command.add("a");
```
Comment (bulasevich) [inline test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java:72]: there is no need in a/b/c params

File: src/hotspot/share/runtime/globals.hpp (hunk 2 above)
Comment (eastig) [reply src/hotspot/share/runtime/globals.hpp]: I have removed it.

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 6]
```diff
@@ -3816,11 +3816,18 @@ encode %{
         ciEnv::current()->record_failure("CodeCache is full");
         return;
       }
-      // Emit stub for static call
-      address stub = CompiledStaticCall::emit_to_interp_stub(cbuf);
-      if (stub == NULL) {
-        ciEnv::current()->record_failure("CodeCache is full");
-        return;
+      if (UseSharedStubs && _method->is_loaded() &&
+          (!_optimized_virtual || _method->is_final_method())) {
```
[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/aarch64.ad]: Now it is:
```
if (CodeBuffer::supports_shared_stubs() && _method->can_be_statically_bound()) {
        // Calls of the same statically bound method can share
        // a stub to the interpreter.
```

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java (hunk 3 above)
Comment (eastig) [reply test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java]: Done

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java (hunk 4 above)
Comment (eastig) [reply test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java]: Done

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java (hunk 5 above)
Comment (eastig) [reply test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java]: Rewritten to remove them.

File: test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                throw new RuntimeException("Found {static_stub} before Deopt Handler Code");
+            } else if (nextLine.contains("{runtime_call DeoptimizationBlob}")) {
+                // Shared static stubs are put after Deopt Handler Code.
+                break;
+            }
+        }
+
+        int foundStaticStubs = 0;
+        while (iter.hasNext()) {
+            if (iter.next().contains("{static_stub}")) {
```
[the remark this answers is not in the document]
Comment (eastig) [reply test/hotspot/jtreg/compiler/sharedstubs/SharedStubToInterpTest.java:119]: No need because `PrintAssembly` is only enabled for one method.

**Comment by eastig** [comment]
Create a new PR after synchronizing with tip: https://github.com/openjdk/jdk/pull/8816

**Pull request closed without merging**